File: src/queue.js
function createQueue() {
  const orders = [];

  function insertionIndex(type) {
    if (type === 'VIP') {
      let i = 0;
      while (i < orders.length && orders[i].type === 'VIP') i++;
      return i;
    }
    return orders.length;
  }

  return {
    enqueue(order) {
      orders.splice(insertionIndex(order.type), 0, order);
    },
    dequeue() {
      return orders.shift();
    },
    returnToQueue(order) {
      let i = 0;
      if (order.type === 'VIP') {
        while (i < orders.length && orders[i].type === 'VIP' && orders[i].id < order.id) i++;
      } else {
        while (i < orders.length && orders[i].type === 'VIP') i++;
        while (i < orders.length && orders[i].id < order.id) i++;
      }
      orders.splice(i, 0, order);
    },
    list() {
      return [...orders];
    },
    get size() {
      return orders.length;
    },
  };
}

module.exports = { createQueue };

File: internal/controller/queue.go
package controller

// PriorityQueue keeps orders in [VIP..., Normal...] arrangement:
// every VIP order is ahead of every Normal order, and orders of the
// same type preserve FIFO arrival. It is a pure data structure —
// synchronization is the caller's (Controller's) responsibility.
type PriorityQueue struct {
	items []*Order
	vips  int // number of VIP orders at the front of items
}

func NewPriorityQueue() *PriorityQueue {
	return &PriorityQueue{}
}

// Enqueue appends a Normal order to the tail, or inserts a VIP order
// right after the last queued VIP (i.e. before the first Normal).
func (q *PriorityQueue) Enqueue(o *Order) {
	if o.Type == VIP {
		q.items = append(q.items, nil)
		copy(q.items[q.vips+1:], q.items[q.vips:])
		q.items[q.vips] = o
		q.vips++
		return
	}
	q.items = append(q.items, o)
}

// Dequeue pops the highest-priority order, or nil when empty.
func (q *PriorityQueue) Dequeue() *Order {
	if len(q.items) == 0 {
		return nil
	}
	o := q.items[0]
	q.items[0] = nil // avoid retaining the popped order
	q.items = q.items[1:]
	if o.Type == VIP {
		q.vips--
	}
	return o
}

func (q *PriorityQueue) Len() int {
	return len(q.items)
}

// Snapshot returns a copy of the current queue contents for reporting.
func (q *PriorityQueue) Snapshot() []*Order {
	out := make([]*Order, len(q.items))
	copy(out, q.items)
	return out
}
